### ISSUE TITLE
Nextjs Middleware doesn't work with webpack URL assets ( `new URL(..., import.meta.url)` ) , `ReferenceError: document is not defined`

### ISSUE THREAD

**Issue by muhajirdev** [vercel/next.js#30646]
### What version of Next.js are you using?

12.0

### What version of Node.js are you using?

v14.17.5

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

vercel

### Describe the Bug

When using middleware, got error `ReferenceError: document is not defined`

Here's the code: https://github.com/quran/quran.com-frontend-next/blob/e64f1200a17ed162a394447898d4f729d8851a36/src/pages/test-geo/_middleware.ts

```
/* eslint-disable import/prefer-default-export */
import type { NextRequest } from 'next/server';

export function middleware(req: NextRequest) {
  const { geo } = req;

  return new Response(JSON.stringify({ message: 'hello world!', geo }), {
    status: 200,
    headers: {
      'Content-Type': 'application/json',
    },
  });
}
````

Here's the API URL: https://quran-com-95pgbz4dx-qurancom.vercel.app/test-geo

And the vercel log is here : https://vercel.com/qurancom/quran-com/5byW9AFxDFa93J9AUtGS97PJR8FV/functions

We tried to pinpoint the issue, turns out the problem is here https://github.com/quran/quran.com-frontend-next/blob/0d120d9e44e590702cd8f58b3a99b110af390eff/src/hooks/useTarteelVoiceSearch.ts#L180

```
new AudioWorklet(new URL('src/audioInput/MicInputProcessor.ts', import.meta.url)),
```

The problem is related to `import.meta.url` https://webpack.js.org/guides/asset-modules/#url-assets

### Expected Behavior

no error

### To Reproduce

The repo is open source
- This is the reference commit

- Install the dependencies
- `yarn dev`
- go to localhost:3000/test-geo

**Comment by TimoWilhelm** [comment]
As the target in the webpack config is `web`, webpack will compile the worker asset module to something like this:

```ts
// target: web
new URL(
  __webpack_public_path__ + 'worker.js',
  document.baseURI || self.location.href
);
```

But both `document` and `location` are undefined in the [sandbox config](https://github.com/javivelasco/next.js/blob/e142cb181564b02583df0362233aa3dce0dc043f/packages/next/server/web/sandbox/sandbox.ts#L41).
Would it be possible to mock these values to allow webpack to resolve the asset modules?

Right now this is unfortunately preventing anyone with workers or asset modules from using middlewares.

**Comment by TimoWilhelm** [comment]
I tried replacing the `__webpack_require__.b` baseUri in the webpack output and it seemed to work fine.

https://github.com/TimoWilhelm/next.js/commit/71174c44bd2c358a710d56cb05c0644119455692

Another option would be to set `document` and `location` to an empty object, but that would probably break code for people assuming the script is executed in a true browser context.

**Comment by TimoWilhelm** [comment]
I tried to reproduce the issue in an integration test using just `const url = new URL('../lib/worker.js', import.meta.url)` as Worker/Worklets are not available in the jsdom environment. but I was unable to reproduce the issue.

I got a different error in the integration test. My best guess is, that it uses a different webpack target.

```
    Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
    TypeError: Cannot convert undefined or null to object
        at getPrototypeOf (<anonymous>)
 ```

```js
      23 | function Index() {
      24 |     const url = /* asset import */ new __webpack_require__.U(__webpack_require__(90));
    > 25 |     console.log(url);
         |             ^
      26 |     return(/*#__PURE__*/ jsx_runtime_namespaceObject.jsx(jsx_runtime_namespaceObject.Fragment, {
      27 |     }));
      28 | };
```

 I also tried to add a E2E test, but I didn't manage to get it running on Windows with swc...

**Comment by javivelasco** [comment]
This was fixed by https://github.com/vercel/next.js/pull/33134 and released in 12.0.8

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
